src/scratchOrgInfoApi.ts

```diff
@@ -0,0 +1,528 @@
+  options: OAuth2Options;
+  retries: number;
+  timeout?: number;
+  delay?: number;
+}> => {
+  const logger = await Logger.child('buildOAuth2Options');
+  const isJwtFlow = !!options.hubOrg.getConnection().getAuthInfoFields().privateKey;
+  const oauth2Options: OAuth2Options = {
+    loginUrl: getOrgInstanceAuthority(
+      options.scratchOrgInfoComplete,
+      options.hubOrg.getField(Org.Fields.LOGIN_URL) as string,
+      options.signupTargetLoginUrlConfig
+    ),
+  };
+
+  logger.debug(`_authorize - isJwtFlow: ${isJwtFlow}`);
+
+  if (isJwtFlow && !process.env.SFDX_CLIENT_SECRET) {
+    oauth2Options.privateKeyFile = options.hubOrg.getConnection().getAuthInfoFields().privateKey;
+    const retries = options?.retry || env.getNumber('SFDX_JWT_AUTH_RETRY_ATTEMPTS') || 0;
+    const timeoutInSeconds = env.getNumber('SFDX_JWT_AUTH_RETRY_TIMEOUT') || 300;
+    const timeout = Duration.seconds(timeoutInSeconds).milliseconds;
+    const delay = retries ? timeout / retries : 1000;
+    return {
+      options: oauth2Options,
+      retries,
+      timeout,
+      delay,
+    };
+  } else {
+    // Web Server OAuth "auth code exchange" flow
+    if (process.env.SFDX_CLIENT_SECRET) {
+      oauth2Options.clientSecret = process.env.SFDX_CLIENT_SECRET;
+    } else if (options.clientSecret) {
+      oauth2Options.clientSecret = options.clientSecret;
+    }
+    oauth2Options.redirectUri = options.scratchOrgInfoComplete.ConnectedAppCallbackUrl;
+    oauth2Options.authCode = options.scratchOrgInfoComplete.AuthCode;
+    return {
+      options: oauth2Options,
+      retries: 0,
+    };
+  }
+};
+
+/**
+ * Returns OAuth2Options object
+ *
+ * @param hubOrg the environment hub org
+ * @param username The OAuth client secret. May be null for JWT OAuth flow.
+ * @param oauth2Options The completed ScratchOrgInfo which should contain an access token.
+ * @param retries auth retry attempts
+ * @param timeout the login url
+ * @param delay the login url
+ * @returns {OAuth2Options, number, number, number} options, retries, timeout, delay
+ */
+const getAuthInfo = async (options: {
+  hubOrg: Org;
+  username: string;
+  oauth2Options: OAuth2Options;
+  retries: number;
+  timeout?: number;
+  delay?: number;
+}): Promise<AuthInfo> => {
+  const logger = await Logger.child('getAuthInfo');
+
+  const retryAuthorize = retryDecorator(async (opts: AuthInfo.Options): Promise<AuthInfo> => AuthInfo.create(opts), {
+    timeout: options.timeout,
+    delay: options.delay,
+    retries: options.retries,
+  });
+
+  if (options.retries) {
+    try {
+      return await retryAuthorize({
+        username: options.username,
+        parentUsername: options.hubOrg.getUsername(),
+        oauth2Options: options.oauth2Options,
+      });
+    } catch (err) {
+      const error = err as RetryError;
+      logger.error(error);
+      if (error.message.startsWith('Timeout after')) {
+        throw SfdxError.create('salesforce-alm', 'org_create', 'jwtAuthRetryTimedOut', [
```

Comment:
This is not correct.  It should be getting an appropriate message from this library.